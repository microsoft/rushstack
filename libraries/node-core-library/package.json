{
  "name": "@rushstack/node-core-library",
  "version": "5.10.1",
  "description": "Core libraries that every NodeJS toolchain project should use",
  "main": "lib/index.js",
  "typings": "dist/node-core-library.d.ts",
  "license": "MIT",
  "repository": {
    "url": "https://github.com/microsoft/rushstack.git",
    "type": "git",
    "directory": "libraries/node-core-library"
  },
  "scripts": {
    "build": "heft build --clean",
    "_phase:build": "heft run --only build -- --clean",
    "_phase:test": "heft run --only test -- --clean"
  },
  "dependencies": {
    "fs-extra": "~7.0.1",
    "import-lazy": "~4.0.0",
    "jju": "~1.4.0",
    "resolve": "~1.22.1",
    "semver": "~7.5.4",
    "ajv": "~8.13.0",
    "ajv-draft-04": "~1.0.0",
    "ajv-formats": "~3.0.1"
  },
  "devDependencies": {
    "local-eslint-config": "workspace:*",
    "@rushstack/heft": "0.68.12",
    "@rushstack/heft-node-rig": "2.6.48",
    "@types/fs-extra": "7.0.0",
    "@types/heft-jest": "1.0.1",
    "@types/jju": "1.4.1",
    "@types/node": "18.17.15",
    "@types/resolve": "1.20.2",
    "@types/semver": "7.5.0"
  },
  "peerDependencies": {
    "@types/node": "*"
  },
  "peerDependenciesMeta": {
    "@types/node": {
      "optional": true
    }
  }
}
