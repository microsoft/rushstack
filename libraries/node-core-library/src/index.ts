// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

/**
 * Core libraries that every NodeJS toolchain project should use.
 *
 * @packageDocumentation
 */

export { AlreadyReportedError } from './AlreadyReportedError';
export {
  Async,
  AsyncQueue,
  type IAsyncParallelismOptions,
  type IRunWithRetriesOptions,
  type IWeighted
} from './Async';
export type { Brand } from './PrimitiveTypes';
export { FileConstants, FolderConstants } from './Constants';
export { Enum } from './Enum';
export { EnvironmentMap, type IEnvironmentEntry } from './EnvironmentMap';
export {
  type ExecutableStdioStreamMapping,
  type ExecutableStdioMapping,
  type IExecutableResolveOptions,
  type IExecutableSpawnSyncOptions,
  type IExecutableSpawnOptions,
  type IWaitForExitOptions,
  type IWaitForExitWithBufferOptions,
  type IWaitForExitWithStringOptions,
  type IWaitForExitResult,
  type IProcessInfo,
  Executable
} from './Executable';
export { type IFileErrorOptions, type IFileErrorFormattingOptions, FileError } from './FileError';
export type {
  INodePackageJson,
  IPackageJson,
  IPackageJsonDependencyTable,
  IPackageJsonScriptTable,
  IPackageJsonRepository,
  IPeerDependenciesMetaTable,
  IDependenciesMetaTable,
  IPackageJsonExports
} from './IPackageJson';
export {
  Import,
  type IImportResolveOptions,
  type IImportResolveAsyncOptions,
  type IImportResolveModuleOptions,
  type IImportResolveModuleAsyncOptions,
  type IImportResolvePackageOptions,
  type IImportResolvePackageAsyncOptions
} from './Import';
export { InternalError } from './InternalError';
export {
  type JsonObject,
  type JsonNull,
  JsonSyntax,
  type IJsonFileParseOptions,
  type IJsonFileLoadAndValidateOptions,
  type IJsonFileStringifyOptions,
  type IJsonFileSaveOptions,
  JsonFile
} from './JsonFile';
export {
  type IJsonSchemaErrorInfo,
  type IJsonSchemaFromFileOptions,
  type IJsonSchemaFromObjectOptions,
  type IJsonSchemaLoadOptions,
  type IJsonSchemaValidateOptions,
  type IJsonSchemaValidateObjectWithOptions,
  JsonSchema,
  type JsonSchemaVersion
} from './JsonSchema';
export { LockFile } from './LockFile';
export { MapExtensions } from './MapExtensions';
export { MinimumHeap } from './MinimumHeap';
export { PosixModeBits } from './PosixModeBits';
export { ProtectableMap, type IProtectableMapParameters } from './ProtectableMap';
export { type IPackageJsonLookupParameters, PackageJsonLookup } from './PackageJsonLookup';
export {
  PackageName,
  PackageNameParser,
  type IPackageNameParserOptions,
  type IParsedPackageName,
  type IParsedPackageNameOrError
} from './PackageName';
export {
  Path,
  type FileLocationStyle,
  type IPathFormatFileLocationOptions,
  type IPathFormatConciselyOptions
} from './Path';
export { Encoding, Text, NewlineKind, type IReadLinesFromIterableOptions } from './Text';
export { Sort } from './Sort';
export {
  AlreadyExistsBehavior,
  FileSystem,
  type FileSystemCopyFilesAsyncFilter,
  type FileSystemCopyFilesFilter,
  type FolderItem,
  type FileSystemStats,
  type IFileSystemCopyFileBaseOptions,
  type IFileSystemCopyFileOptions,
  type IFileSystemCopyFilesAsyncOptions,
  type IFileSystemCopyFilesOptions,
  type IFileSystemCreateLinkOptions,
  type IFileSystemDeleteFileOptions,
  type IFileSystemMoveOptions,
  type IFileSystemReadFileOptions,
  type IFileSystemReadFolderOptions,
  type IFileSystemUpdateTimeParameters,
  type IFileSystemWriteBinaryFileOptions,
  type IFileSystemWriteFileOptions
} from './FileSystem';
export { FileWriter, type IFileWriterFlags } from './FileWriter';
export { LegacyAdapters, type LegacyCallback } from './LegacyAdapters';
export { StringBuilder, type IStringBuilder } from './StringBuilder';
export { type ISubprocessOptions, SubprocessTerminator } from './SubprocessTerminator';
export { TypeUuid } from './TypeUuid';
