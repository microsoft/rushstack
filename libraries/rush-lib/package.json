{
  "name": "@microsoft/rush-lib",
  "version": "5.136.1",
  "description": "A library for writing scripts that interact with the Rush tool",
  "repository": {
    "type": "git",
    "url": "https://github.com/microsoft/rushstack.git",
    "directory": "libraries/rush-lib"
  },
  "engines": {
    "node": ">=5.6.0"
  },
  "engineStrict": true,
  "homepage": "https://rushjs.io",
  "main": "lib/index.js",
  "typings": "dist/rush-lib.d.ts",
  "typesVersions": {
    "*": {
      "lib-esnext/*": [
        "lib/*"
      ]
    }
  },
  "scripts": {
    "build": "heft build --clean",
    "test": "heft test --clean",
    "_phase:build": "heft run --only build -- --clean",
    "_phase:test": "heft run --only test -- --clean"
  },
  "license": "MIT",
  "dependencies": {
    "@pnpm/dependency-path": "~2.1.2",
    "@pnpm/link-bins": "~5.3.7",
    "@rushstack/heft-config-file": "workspace:*",
    "@rushstack/lookup-by-path": "workspace:*",
    "@rushstack/node-core-library": "workspace:*",
    "@rushstack/package-deps-hash": "workspace:*",
    "@rushstack/package-extractor": "workspace:*",
    "@rushstack/rig-package": "workspace:*",
    "@rushstack/stream-collator": "workspace:*",
    "@rushstack/terminal": "workspace:*",
    "@rushstack/ts-command-line": "workspace:*",
    "@types/node-fetch": "2.6.2",
    "@yarnpkg/lockfile": "~1.0.2",
    "builtin-modules": "~3.1.0",
    "cli-table": "~0.3.1",
    "dependency-path": "~9.2.8",
    "fast-glob": "~3.3.1",
    "figures": "3.0.0",
    "git-repo-info": "~2.1.0",
    "glob-escape": "~0.0.2",
    "https-proxy-agent": "~5.0.0",
    "ignore": "~5.1.6",
    "inquirer": "~7.3.3",
    "js-yaml": "~3.13.1",
    "node-fetch": "2.6.7",
    "npm-check": "~6.0.1",
    "npm-package-arg": "~6.1.0",
    "read-package-tree": "~5.1.5",
    "rxjs": "~6.6.7",
    "semver": "~7.5.4",
    "ssri": "~8.0.0",
    "strict-uri-encode": "~2.0.0",
    "tapable": "2.2.1",
    "tar": "~6.2.1",
    "true-case-path": "~2.2.1",
    "uuid": "~8.3.2",
    "pnpm-sync-lib": "0.2.9"
  },
  "devDependencies": {
    "@pnpm/logger": "4.0.0",
    "local-node-rig": "workspace:*",
    "@rushstack/heft-webpack5-plugin": "workspace:*",
    "@rushstack/heft": "workspace:*",
    "@rushstack/operation-graph": "workspace:*",
    "@rushstack/webpack-deep-imports-plugin": "workspace:*",
    "@rushstack/webpack-preserve-dynamic-require-plugin": "workspace:*",
    "@types/cli-table": "0.3.0",
    "@types/inquirer": "7.3.1",
    "@types/js-yaml": "3.12.1",
    "@types/npm-package-arg": "6.1.0",
    "@types/read-package-tree": "5.1.0",
    "@types/semver": "7.5.0",
    "@types/ssri": "~7.1.0",
    "@types/strict-uri-encode": "2.0.0",
    "@types/tar": "6.1.6",
    "@types/uuid": "~8.3.4",
    "@types/webpack-env": "1.18.0",
    "webpack": "~5.95.0"
  },
  "publishOnlyDependencies": {
    "@rushstack/rush-amazon-s3-build-cache-plugin": "workspace:*",
    "@rushstack/rush-azure-storage-build-cache-plugin": "workspace:*",
    "@rushstack/rush-http-build-cache-plugin": "workspace:*"
  },
  "sideEffects": [
    "lib-esnext/start-pnpm.js",
    "lib-esnext/start.js",
    "lib-esnext/startx.js",
    "lib-esnext/utilities/SetRushLibPath.js"
  ]
}
