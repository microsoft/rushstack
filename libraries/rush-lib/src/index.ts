// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

/**
 * A library for writing scripts that interact with the {@link https://rushjs.io/ | Rush} tool.
 * @packageDocumentation
 */

// For backwards compatibility
export { LookupByPath as LookupByPath, type IPrefixMatch } from '@rushstack/lookup-by-path';

export { ApprovedPackagesPolicy } from './api/ApprovedPackagesPolicy';

export { RushConfiguration, type ITryFindRushJsonLocationOptions } from './api/RushConfiguration';

export { Subspace } from './api/Subspace';
export { SubspacesConfiguration } from './api/SubspacesConfiguration';

export {
  type IPackageManagerOptionsJsonBase,
  type IConfigurationEnvironment,
  type IConfigurationEnvironmentVariable,
  PackageManagerOptionsConfigurationBase
} from './logic/base/BasePackageManagerOptionsConfiguration';
export {
  type INpmOptionsJson as _INpmOptionsJson,
  NpmOptionsConfiguration
} from './logic/npm/NpmOptionsConfiguration';
export {
  type IYarnOptionsJson as _IYarnOptionsJson,
  YarnOptionsConfiguration
} from './logic/yarn/YarnOptionsConfiguration';
export {
  type IPnpmOptionsJson as _IPnpmOptionsJson,
  type PnpmStoreLocation,
  type IPnpmLockfilePolicies,
  type IPnpmPackageExtension,
  type IPnpmPeerDependencyRules,
  type IPnpmPeerDependenciesMeta,
  type PnpmStoreOptions,
  PnpmOptionsConfiguration,
  type PnpmResolutionMode
} from './logic/pnpm/PnpmOptionsConfiguration';

export { BuildCacheConfiguration } from './api/BuildCacheConfiguration';
export { CobuildConfiguration, type ICobuildJson } from './api/CobuildConfiguration';
export type { GetCacheEntryIdFunction, IGenerateCacheEntryIdOptions } from './logic/buildCache/CacheEntryId';
export {
  FileSystemBuildCacheProvider,
  type IFileSystemBuildCacheProviderOptions
} from './logic/buildCache/FileSystemBuildCacheProvider';

export type {
  IPhase,
  PhaseBehaviorForMissingScript as IPhaseBehaviorForMissingScript
} from './api/CommandLineConfiguration';

export {
  EnvironmentConfiguration,
  EnvironmentVariableNames,
  type IEnvironmentConfigurationInitializeOptions
} from './api/EnvironmentConfiguration';

export { RushConstants } from './logic/RushConstants';

export { type PackageManagerName, PackageManager } from './api/packageManager/PackageManager';

export { RushConfigurationProject } from './api/RushConfigurationProject';

export {
  type IRushProjectJson as _IRushProjectJson,
  type IOperationSettings,
  RushProjectConfiguration,
  type IRushPhaseSharding
} from './api/RushProjectConfiguration';

export { RushUserConfiguration } from './api/RushUserConfiguration';

export { RushGlobalFolder as _RushGlobalFolder } from './api/RushGlobalFolder';

export { ApprovedPackagesItem, ApprovedPackagesConfiguration } from './api/ApprovedPackagesConfiguration';

export { CommonVersionsConfiguration } from './api/CommonVersionsConfiguration';

export {
  PackageJsonEditor,
  PackageJsonDependency,
  DependencyType,
  PackageJsonDependencyMeta
} from './api/PackageJsonEditor';

export { RepoStateFile } from './logic/RepoStateFile';

export { EventHooks, Event } from './api/EventHooks';

export { ChangeManager } from './api/ChangeManager';

export { FlagFile as _FlagFile } from './api/FlagFile';

export {
  VersionPolicyDefinitionName,
  BumpType,
  LockStepVersionPolicy,
  IndividualVersionPolicy,
  VersionPolicy
} from './api/VersionPolicy';

export { VersionPolicyConfiguration } from './api/VersionPolicyConfiguration';

export { type ILaunchOptions, Rush } from './api/Rush';
export { RushInternals as _RushInternals } from './api/RushInternals';

export { ExperimentsConfiguration, type IExperimentsJson } from './api/ExperimentsConfiguration';
export {
  CustomTipsConfiguration,
  CustomTipId,
  type ICustomTipsJson,
  type ICustomTipInfo,
  type ICustomTipItemJson,
  CustomTipSeverity,
  CustomTipType
} from './api/CustomTipsConfiguration';

export { ProjectChangeAnalyzer, type IGetChangedProjectsOptions } from './logic/ProjectChangeAnalyzer';
export type {
  IInputsSnapshot,
  GetInputsSnapshotAsyncFn as GetInputsSnapshotAsyncFn,
  IRushConfigurationProjectForSnapshot
} from './logic/incremental/InputsSnapshot';

export type { IOperationRunner, IOperationRunnerContext } from './logic/operations/IOperationRunner';
export type {
  IExecutionResult,
  IOperationExecutionResult
} from './logic/operations/IOperationExecutionResult';
export { type IOperationOptions, Operation } from './logic/operations/Operation';
export { OperationStatus } from './logic/operations/OperationStatus';
export type { ILogFilePaths } from './logic/operations/ProjectLogWritable';

export {
  RushSession,
  type IRushSessionOptions,
  type CloudBuildCacheProviderFactory,
  type CobuildLockProviderFactory
} from './pluginFramework/RushSession';

export {
  type IRushCommand,
  type IGlobalCommand,
  type IPhasedCommand,
  RushLifecycleHooks
} from './pluginFramework/RushLifeCycle';

export {
  type ICreateOperationsContext,
  type IExecuteOperationsContext,
  PhasedCommandHooks
} from './pluginFramework/PhasedCommandHooks';

export type { IRushPlugin } from './pluginFramework/IRushPlugin';
export type { IBuiltInPluginConfiguration as _IBuiltInPluginConfiguration } from './pluginFramework/PluginLoader/BuiltInPluginLoader';
export type { IRushPluginConfigurationBase as _IRushPluginConfigurationBase } from './api/RushPluginsConfiguration';
export type { ILogger } from './pluginFramework/logging/Logger';

export type { ICloudBuildCacheProvider } from './logic/buildCache/ICloudBuildCacheProvider';
export type {
  ICobuildLockProvider,
  ICobuildContext,
  ICobuildCompletedState
} from './logic/cobuild/ICobuildLockProvider';

export {
  type ICredentialCacheOptions,
  type ICredentialCacheEntry,
  CredentialCache
} from './logic/CredentialCache';

export type { ITelemetryData, ITelemetryMachineInfo, ITelemetryOperationResult } from './logic/Telemetry';

export type { IStopwatchResult } from './utilities/Stopwatch';
export {
  OperationStateFile as _OperationStateFile,
  type IOperationStateFileOptions as _IOperationStateFileOptions,
  type IOperationStateJson as _IOperationStateJson
} from './logic/operations/OperationStateFile';
export {
  OperationMetadataManager as _OperationMetadataManager,
  type IOperationMetadataManagerOptions as _IOperationMetadataManagerOptions,
  type IOperationMetaData as _IOperationMetadata
} from './logic/operations/OperationMetadataManager';

export {
  RushCommandLine,
  type IRushCommandLineSpec,
  type IRushCommandLineParameter,
  type IRushCommandLineAction
} from './api/RushCommandLine';
