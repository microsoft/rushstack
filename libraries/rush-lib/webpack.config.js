'use strict';

const webpack = require('webpack');
const { PackageJsonLookup } = require('@rushstack/node-core-library');
const { PreserveDynamicRequireWebpackPlugin } = require('@rushstack/webpack-preserve-dynamic-require-plugin');
const { DeepImportsPlugin } = require('@rushstack/webpack-deep-imports-plugin');
const PathConstants = require('./lib-commonjs/utilities/PathConstants');

const SCRIPT_ENTRY_OPTIONS = {
  filename: `${PathConstants.scriptsFolderName}/[name]`
};

module.exports = () => {
  const packageJson = PackageJsonLookup.loadOwnPackageJson(__dirname);

  const externalDependencyNames = new Set([
    ...Object.keys(packageJson.dependencies || {}),
    ...Object.keys(packageJson.peerDependencies || {}),
    ...Object.keys(packageJson.optionalDependencies || {}),
    ...Object.keys(packageJson.devDependencies || {})
  ]);

  function generateConfiguration(entry, extraPlugins = [], splitChunks = undefined) {
    return {
      context: __dirname,
      mode: 'development', // So the output isn't minified
      devtool: 'source-map',
      entry,
      output: {
        path: `${__dirname}/dist`,
        filename: '[name].js',
        chunkFilename: 'chunks/[name].js',
        library: {
          type: 'commonjs2'
        }
      },
      target: 'node',
      plugins: [
        new PreserveDynamicRequireWebpackPlugin(),
        new webpack.ids.DeterministicModuleIdsPlugin({
          maxLength: 6
        }),
        ...extraPlugins
      ],
      externals: [
        ({ request }, callback) => {
          let packageName;
          let firstSlashIndex = request.indexOf('/');
          if (firstSlashIndex === -1) {
            packageName = request;
          } else if (request.startsWith('@')) {
            let secondSlash = request.indexOf('/', firstSlashIndex + 1);
            if (secondSlash === -1) {
              packageName = request;
            } else {
              packageName = request.substring(0, secondSlash);
            }
          } else {
            packageName = request.substring(0, firstSlashIndex);
          }

          if (externalDependencyNames.has(packageName)) {
            callback(null, `commonjs ${request}`);
          } else {
            callback();
          }
        }
      ],
      optimization: {
        splitChunks
      }
    };
  }

  const configurations = [
    generateConfiguration(
      {
        'rush-lib': `${__dirname}/lib-esnext/index.js`,
        start: `${__dirname}/lib-esnext/start.js`,
        startx: `${__dirname}/lib-esnext/startx.js`,
        'start-pnpm': `${__dirname}/lib-esnext/start-pnpm.js`
      },
      [
        new DeepImportsPlugin({
          // A manifest will be produced for each entry point, so since this compilation has multiple entry points,
          // it needs to specify a template for the manifest filename.
          // Otherwise webpack will throw an error about multiple writes to the same manifest file.
          path: `${__dirname}/temp/build/webpack-dll/[name].json`,
          inFolderName: 'lib-esnext',
          outFolderName: 'lib',
          pathsToIgnore: ['utilities/prompts/SearchListPrompt.js'],
          dTsFilesInputFolderName: 'lib-commonjs'
        })
      ],
      {
        chunks: 'all',
        minChunks: 1,
        cacheGroups: {
          commons: {
            name: 'commons',
            chunks: 'initial',
            minChunks: 2
          }
        }
      }
    ),
    generateConfiguration({
      [PathConstants.pnpmfileShimFilename]: {
        import: `${__dirname}/lib-esnext/logic/pnpm/PnpmfileShim.js`,
        ...SCRIPT_ENTRY_OPTIONS
      },
      [PathConstants.subspacePnpmfileShimFilename]: {
        import: `${__dirname}/lib-esnext/logic/pnpm/SubspaceGlobalPnpmfileShim.js`,
        ...SCRIPT_ENTRY_OPTIONS
      },
      [PathConstants.installRunScriptFilename]: {
        import: `${__dirname}/lib-esnext/scripts/install-run.js`,
        ...SCRIPT_ENTRY_OPTIONS
      },
      [PathConstants.installRunRushScriptFilename]: {
        import: `${__dirname}/lib-esnext/scripts/install-run-rush.js`,
        ...SCRIPT_ENTRY_OPTIONS
      },
      [PathConstants.installRunRushxScriptFilename]: {
        import: `${__dirname}/lib-esnext/scripts/install-run-rushx.js`,
        ...SCRIPT_ENTRY_OPTIONS
      },
      [PathConstants.installRunRushPnpmScriptFilename]: {
        import: `${__dirname}/lib-esnext/scripts/install-run-rush-pnpm.js`,
        ...SCRIPT_ENTRY_OPTIONS
      }
    })
  ];

  return configurations;
};
