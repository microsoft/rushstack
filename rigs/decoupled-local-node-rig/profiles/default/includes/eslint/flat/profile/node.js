// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

// This profile enables lint rules intended for a general Node.js project, typically a web service.
// It enables security rules that assume the service could receive malicious inputs from an
// untrusted user.  If that is not the case, consider using the "node-trusted-tool" profile instead.

const nodeProfile = require('@rushstack/eslint-config/flat/profile/node');

const { localCommonConfig } = require('./_common');

module.exports = [...nodeProfile, ...localCommonConfig];
