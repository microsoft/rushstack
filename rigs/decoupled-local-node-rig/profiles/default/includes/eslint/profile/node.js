// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

// This profile enables lint rules intended for a general Node.js project, typically a web service.
// It enables security rules that assume the service could receive malicious inputs from an
// untrusted user.  If that is not the case, consider using the "node-trusted-tool" profile instead.

const { buildRules } = require('./_common');

const rules = buildRules('node');
module.exports = rules;
