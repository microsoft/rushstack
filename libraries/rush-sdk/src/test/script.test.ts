// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import * as path from 'node:path';
import { Executable } from '@rushstack/node-core-library';

const rushSdkPath: string = path.join(__dirname, '../../lib-shim/index.js');
const sandboxRepoPath: string = `${__dirname}/sandbox`;
const mockPackageFolder: string = `${sandboxRepoPath}/mock-package`;
const mockRushLibPath: string = `${__dirname}/fixture/mock-rush-lib.js`;

const coreLibPath: string = require.resolve('@rushstack/node-core-library');
const quotedRushSdkPath: string = JSON.stringify(rushSdkPath);
const loadAndPrintRushSdkModule: string = `console.log(JSON.stringify(Object.keys(require(${quotedRushSdkPath})).sort(), undefined, 2).replace(/"/g, "'"));`;

describe('@rushstack/rush-sdk', () => {
  it('Should load via global (for plugins)', () => {
    const result = Executable.spawnSync(
      'node',
      [
        '-e',
        `
global.___rush___rushLibModule = { foo: 1 };
${loadAndPrintRushSdkModule}`
      ],
      {
        currentWorkingDirectory: mockPackageFolder,
        environment: {
          ...process.env,
          RUSH_SDK_DEBUG: '1',
          _RUSH_LIB_PATH: '' // Need to clear if invoked via Rush
        }
      }
    );
    expect(result.stderr.trim()).toMatchSnapshot('stderr');
    expect(result.stdout.trim()).toMatchSnapshot('stdout');
    expect(result.status).toBe(0);
  });

  it('Should load via env when Rush has loaded (for child processes)', () => {
    const result = Executable.spawnSync(
      'node',
      [
        '-e',
        `
require('@microsoft/rush-lib');
${loadAndPrintRushSdkModule}`
      ],
      {
        currentWorkingDirectory: mockPackageFolder,
        environment: {
          ...process.env,
          RUSH_SDK_DEBUG: '1',
          _RUSH_LIB_PATH: '' // Need to clear if invoked via Rush
        }
      }
    );
    expect(result.stderr.trim()).toMatchSnapshot('stderr');
    expect(result.stdout.trim()).toMatchSnapshot('stdout');
    expect(result.status).toBe(0);
  });

  it('Should load via process.env._RUSH_LIB_PATH (for child processes)', () => {
    const result = Executable.spawnSync('node', ['-e', loadAndPrintRushSdkModule], {
      currentWorkingDirectory: mockPackageFolder,
      environment: {
        ...process.env,
        RUSH_SDK_DEBUG: '1',
        _RUSH_LIB_PATH: mockRushLibPath
      }
    });
    expect(result.stderr.trim()).toMatchSnapshot('stderr');
    expect(result.stdout.trim()).toMatchSnapshot('stdout');
    expect(result.status).toBe(0);
  });

  it('Should load via install-run (for standalone tools)', () => {
    const result = Executable.spawnSync(
      'node',
      [
        '-e',
        `
const { Import } = require(${JSON.stringify(coreLibPath)});
const originalResolveModule = Import.resolveModule;
const mockResolveModule = (options) => {
  if (options.baseFolderPath.includes('install-run') && options.modulePath === '@microsoft/rush-lib') {
    return ${JSON.stringify(mockRushLibPath)};
  }
  return originalResolveModule(options);
}
Import.resolveModule = mockResolveModule;
${loadAndPrintRushSdkModule}
`
      ],
      {
        currentWorkingDirectory: mockPackageFolder,
        environment: {
          ...process.env,
          RUSH_SDK_DEBUG: '1',
          _RUSH_LIB_PATH: '' // Need to clear if invoked via Rush
        }
      }
    );
    expect(result.stderr.trim()).toMatchSnapshot('stderr');
    expect(result.stdout.trim()).toMatchSnapshot('stdout');
    expect(result.status).toBe(0);
  });
});
