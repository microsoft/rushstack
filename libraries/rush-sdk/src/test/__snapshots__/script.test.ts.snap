// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@rushstack/rush-sdk Should load via env when Rush has loaded (for child processes): stderr 1`] = `""`;

exports[`@rushstack/rush-sdk Should load via env when Rush has loaded (for child processes): stdout 1`] = `
"Try to load @microsoft/rush-lib from process.env._RUSH_LIB_PATH from caller package
Loaded @microsoft/rush-lib from process.env._RUSH_LIB_PATH
[
  'ApprovedPackagesConfiguration',
  'ApprovedPackagesItem',
  'ApprovedPackagesPolicy',
  'BuildCacheConfiguration',
  'BumpType',
  'ChangeManager',
  'CobuildConfiguration',
  'CommonVersionsConfiguration',
  'CredentialCache',
  'CustomTipId',
  'CustomTipSeverity',
  'CustomTipType',
  'CustomTipsConfiguration',
  'DependencyType',
  'EnvironmentConfiguration',
  'EnvironmentVariableNames',
  'Event',
  'EventHooks',
  'ExperimentsConfiguration',
  'FileSystemBuildCacheProvider',
  'IndividualVersionPolicy',
  'LockStepVersionPolicy',
  'LookupByPath',
  'NpmOptionsConfiguration',
  'Operation',
  'OperationStatus',
  'PackageJsonDependency',
  'PackageJsonDependencyMeta',
  'PackageJsonEditor',
  'PackageManager',
  'PackageManagerOptionsConfigurationBase',
  'PhasedCommandHooks',
  'PnpmOptionsConfiguration',
  'ProjectChangeAnalyzer',
  'RepoStateFile',
  'Rush',
  'RushCommandLine',
  'RushConfiguration',
  'RushConfigurationProject',
  'RushConstants',
  'RushLifecycleHooks',
  'RushProjectConfiguration',
  'RushSession',
  'RushUserConfiguration',
  'Subspace',
  'SubspacesConfiguration',
  'VersionPolicy',
  'VersionPolicyConfiguration',
  'VersionPolicyDefinitionName',
  'YarnOptionsConfiguration',
  '_FlagFile',
  '_OperationMetadataManager',
  '_OperationStateFile',
  '_RushGlobalFolder',
  '_RushInternals',
  '_rushSdk_loadInternalModule'
]"
`;

exports[`@rushstack/rush-sdk Should load via global (for plugins): stderr 1`] = `""`;

exports[`@rushstack/rush-sdk Should load via global (for plugins): stdout 1`] = `
"[
  '_rushSdk_loadInternalModule',
  'foo'
]"
`;

exports[`@rushstack/rush-sdk Should load via install-run (for standalone tools): stderr 1`] = `""`;

exports[`@rushstack/rush-sdk Should load via install-run (for standalone tools): stdout 1`] = `
"Try to load @microsoft/rush-lib from rush global folder
The expected global rush installed folder is \\"/Users/ianc/.rush/node-v22.12.0/rush-5.57.0\\"
Failed to load @microsoft/rush-lib from rush global folder: File does not exist: /Users/ianc/.rush/node-v22.12.0/rush-5.57.0
ENOENT: no such file or directory, lstat '/Users/ianc/.rush/node-v22.12.0/rush-5.57.0'
Trying to load  @microsoft/rush-lib installed by install-run-rush
Loaded @microsoft/rush-lib installed by install-run-rush
[
  '_rushSdk_loadInternalModule',
  'foo'
]"
`;

exports[`@rushstack/rush-sdk Should load via process.env._RUSH_LIB_PATH (for child processes): stderr 1`] = `""`;

exports[`@rushstack/rush-sdk Should load via process.env._RUSH_LIB_PATH (for child processes): stdout 1`] = `
"Try to load @microsoft/rush-lib from process.env._RUSH_LIB_PATH from caller package
Loaded @microsoft/rush-lib from process.env._RUSH_LIB_PATH
[
  '_rushSdk_loadInternalModule',
  'foo'
]"
`;
